### PR TITLE
Closes Issue #2153 -  Limit how a user becomes an admin #2153 

### LINKED ISSUE #2153: Limit how a user becomes an admin
We have the concept of an `admin` role in the User and Auth services, in the front-end, etc.  In Firebase it's a boolean, `isAdmin: true|false`.  We need to tighten how a user becomes an admin.

You shouldn't be able to POST that data to the Users service (we should ignore it).   The current system is too loose.

### PR DESCRIPTION
**Pull request by chrispinkney**
Co-authored-by: Josue manekenpix@fastmail.com

## Issue This PR Addresses
Closes #2153

## Type of Change

- [x] **Bugfix**: Change which fixes an issue
- [x] **New Feature**: Change which adds functionality
- [ ] **Documentation Update**: Change which improves documentation
- [ ] **UI**: Change which improves UI

## Description
This PR limits how a user can become an admin:
- It does this by forcing the `PUT` and `POST` routes to set `user.isAdmin` to `false`.
- It also adds another route for admins to modify another user's `isAdmin`.

## Checklist

- [ ] **Quality**: This PR builds and passes our npm test and works locally
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not (if applicable)
- [ ] **Documentation**: This PR includes updated/added documentation to user exposed functionality or configuration variables are added/changed or an explanation of why it does not(if applicable)

### CODE REVIEW

File: src/api/users/src/routes/users.js [hunk 1]
```diff
@@ -111,6 +111,8 @@ router.post(
         next(createError(409, `user with id ${id} already exists.`));
       } else {
         const user = new User(body);
+        // no user can be created as an admin by default
+        user.isAdmin = false;
```
Comment (humphd) [inline src/api/users/src/routes/users.js:115]: I would move this code into a new middleware function that happens just before the route's logic gets hit.  Maybe call it `validateUserRights()` and it can simply set `user.isAdmin` to `false`.

The logic in the route should be focused on talking to the db and returning HTTP results, nothing more.

File: src/api/users/src/routes/users.js [hunk 2]
```diff
@@ -140,6 +142,43 @@ router.put(
...
 
+    try {
+      const userRef = db.doc(id);
+      const doc = await userRef.get();
+
+      if (!doc.exists) {
+        next(createError(404, `user ${id} not found.`));
+      } else {
+        const user = new User(body);
+        // no user can be created as an admin by default
```
Comment (humphd) [inline src/api/users/src/routes/users.js:153]: Same, this can be done via middleware.

File: src/api/users/src/routes/users.js [hunk 3]
```diff
@@ -140,6 +142,43 @@ router.put(
     const { id } = req.params;
     const { body } = req;
 
+    try {
+      const userRef = db.doc(id);
+      const doc = await userRef.get();
+
+      if (!doc.exists) {
```
Comment (humphd) [inline src/api/users/src/routes/users.js:149]: This code seems to be duplicated between a few `put` calls now.  Can you extract it into middleware and share it?

File: src/api/users/src/routes/users.js [hunk 4]
```diff
@@ -140,6 +142,43 @@ router.put(
...
+    }
+  }
+);
+
+// put (update) a user with a supplied id, validated by the schema
+// rejected if a user could not be found, otherwise user updated
+// this route is only accessible by administrators
+// it allows the modification of the isAdmin property
+router.put(
+  '/:id/admin',
```
Comment (humphd) [inline src/api/users/src/routes/users.js:170]: This looks great, but one question: the naming implies that you are only setting the `admin` aspect of this record vs. all fields, yet you are allowing other fields to be updated here.  I would limit it to just the admin change.

File: src/api/users/src/routes/users.js [hunk 5]
```diff
@@ -140,6 +142,43 @@ router.put(
...
+  isAuthenticated(),
+  validateId(),
+  validateUser(),
+  validateEmailHash(),
+  isAuthorized((req, user) => {
```
Comment (humphd) [inline src/api/users/src/routes/users.js:175]: Arrow functions with body statements that immediately return a value should be rewritten as inline expressions:

```js
isAuthorized((req, user) => user.roles.includes('admin')),
```

**Review by humphd — CHANGES_REQUESTED** [review]
This also lacks tests, which I'd like to see get added with these changes.

**Comment by PedroFonsecaDEV** [comment]
Following.

File: src/api/users/src/routes/users.js (hunk 1 above)
Comment (chrispinkney) [reply src/api/users/src/routes/users.js]: `validateUserRights` created.

File: src/api/users/src/routes/users.js (hunk 2 above)
Comment (chrispinkney) [reply src/api/users/src/routes/users.js]: Done.

File: src/api/users/src/routes/users.js (hunk 3 above)
Comment (chrispinkney) [reply src/api/users/src/routes/users.js]: Duplication removed, code moved to `updateUser` in `schema.js`

File: src/api/users/src/routes/users.js (hunk 4 above)
Comment (chrispinkney) [reply src/api/users/src/routes/users.js:151]: Also done in `updateUser`

File: src/api/users/src/routes/users.js (hunk 5 above)
Comment (chrispinkney) [reply src/api/users/src/routes/users.js]: Done

**Comment by chrispinkney** [comment]
Still have to add tests but I think this is heading in the right direction.

File: config/export/firebase-export-metadata.json [hunk 6]
```diff
@@ -0,0 +1,8 @@
+{
```
Comment (humphd) [inline config/export/firebase-export-metadata.json:1]: Why did this file, and the other `config/export/*` files, get added?

File: docker/development.yml [hunk 7]
```diff
@@ -45,7 +45,8 @@ services:
     volumes:
       # Copy firebase.config into the container so we get proper ip/port binding
       - ../config/firebase.json:/home/node/firebase.json
-    command: firebase emulators:start --project telescope --only firestore
+      - ../config/export:/export
+    command: firebase emulators:start --project telescope --only firestore --import=/export
```
Comment (humphd) [inline docker/development.yml:49]: Why this change?

File: src/api/users/src/models/schema.js [hunk 8]
```diff
@@ -16,6 +16,34 @@ const validateId = () =>
...
+
+  try {
+    const userRef = db.doc(id);
+    const doc = await userRef.get();
+
+    if (!doc.exists) {
+      next(createError(404, `user ${id} not found.`));
+    } else {
+      // if makeAdmin is true turn the user into an admin, if not use body to update user.
+      const user = new User(makeAdmin ? { ...body, isAdmin: true } : body);
```
Comment (humphd) [inline src/api/users/src/models/schema.js:37]: This won't allow downgrading a user from Admin to normal, which we should support.  We should accept a value `{isAdmin: true|false}` on the body and use that instead of hard-coding here.

**Review by humphd — CHANGES_REQUESTED** [review]

File: config/export/firebase-export-metadata.json (hunk 6 above)
Comment (chrispinkney) [reply config/export/firebase-export-metadata.json:1]: This and the change to `development.yml` add a single user to the emulator up its bootup. This user as the `isAdmin` flag set to true, as without this we'd have no way to make a user an admin during testing.

The problem stems from having no way of creating an admin without first having an admin to create an admin.

**Comment by chrispinkney** [comment]
Given #2555, I wonder if this PR should be closed and migrated to a different issue at a later date (once the suprabase work has been merged)?

@manekenpix
@humphd

**Comment by humphd** [comment]
Agree, this can get closed for now; though, I think it's a good starting point for what needs to happen.
